core/shared/src/main/scala/fs2/Segment.scala

```diff
@@ -333,11 +333,12 @@ abstract class Segment[+O,+R] { self =>
 
       outerStep.map { outer =>
         step {
+          val innerRem = if (inner eq null) Segment.empty[O2].asInstanceOf[Segment[O2,S]] else inner.remainder
```

Comment:
Isolated to just this line, this cast is not safe. The `innerRem` val is used 3 places below -- the two `prepend(innerRem)` cases are safe though, as we will never look at the result of `innerRem`. The other use case is not safe though.